src/tls_relation.py

```diff
@@ -5,35 +5,54 @@
 """Gateway API TLS relation business logic."""
 import secrets
 import string
-from typing import Dict, Union
+import typing
 
 from charms.tls_certificates_interface.v3.tls_certificates import (
-    CertificateAvailableEvent,
     CertificateExpiringEvent,
     CertificateInvalidatedEvent,
+    ProviderCertificate,
     TLSCertificatesRequiresV3,
     generate_csr,
     generate_private_key,
 )
 from cryptography import x509
 from cryptography.x509.oid import NameOID
-from ops.jujuversion import JujuVersion
-from ops.model import Model, Relation, SecretNotFoundError
+from ops.model import Relation, SecretNotFoundError
 
 TLS_CERT = "certificates"
 
 
+def get_hostname_from_cert(certificate: str) -> str:
+    """Get the hostname from a certificate subject name.
+
+    Args:
+        certificate: The certificate in PEM format.
+
+    Returns:
+        The hostname the certificate is issue to.
+    """
+    decoded_cert = x509.load_pem_x509_certificate(certificate.encode())
+
+    common_name_attribute = decoded_cert.subject.get_attributes_for_oid(NameOID.COMMON_NAME)
+    if not common_name_attribute:
+        return ""
+
+    return str(common_name_attribute[0].value)
+
+
 class TLSRelationService:
     """TLS Relation service class."""
 
-    def __init__(self, model: Model) -> None:
+    def __init__(self, certificates: TLSCertificatesRequiresV3) -> None:
         """Init method for the class.
 
         Args:
-            model: The charm model used to get the relations and secrets.
+            certificates: The TLS certificates requirer library.
         """
-        self.charm_model = model
-        self.charm_app = model.app
+        self.certificates = certificates
+        self.model = self.certificates.model
```

Comment:
It's a weird way to access the model, very indirect. Why removing the model from the constructor ?

Reply:
I was aiming to reduce the number of dependencies for the initialization of this module. I added back the `model` parameter.